engines/sci/engine/kgraphics.cpp

```diff
@@ -328,7 +329,9 @@ reg_t kGraphAdjustPriority(EngineState *s, int argc, reg_t *argv) {
 
 reg_t kGraphSaveUpscaledHiresBox(EngineState *s, int argc, reg_t *argv) {
 	Common::Rect rect = getGraphRect(argv);
-	return g_sci->_gfxPaint16->kernelGraphSaveUpscaledHiresBox(rect);
+	uint16 screenMask = argv[4].toUint16() & GFX_SCREEN_MASK_ALL;
+	s->r_acc = g_sci->_gfxScreen->gfxDriver()->supportsHiResGraphics() ? g_sci->_gfxPaint16->kernelGraphSaveBox(rect, screenMask, true) : NULL_REG;
```

Comment:
No need for `s->r_acc`, this can just be:

```
if (g_sci->_gfxScreen->gfxDriver()->supportsHiResGraphics())
	return g_sci->_gfxPaint16->kernelGraphSaveBox(rect, screenMask, true);
else
	return NULL_REG;
```
	
Our kernel function return values are always assigned to the accumulator (unlike the original) so `s->r_acc` only appears in our functions when we're not setting the accumulator, by using the previous value.

Reply:
Also, what's going on with argv[4] here? I don't see anywhere in the game script that calls kGraph 15 with that many parameters. `getGraphRect` reads a rect from argv[0 - 3], so where's this extra param coming from?

I'm only seeing calls that look like this: https://github.com/sluicebox/sci-scripts/blob/2f0970201770878802c2c8b870c8e5b9a8c93336/kq6-cd-dos-1.000.00G/src/Kq6IconBar.sc#L74

Reply:
In the original interpreter it is called the same way as kGraph case 7.


![Screenshot 2024-10-07 225432](https://github.com/user-attachments/assets/3016ebda-f0a6-4e94-9843-447fe9169212)

Reply:
Weird. I tried the village scene with the lamp seller. We seem to get a argc counter of 4 (so argv[4] really is out ouf bounds), but nonetheless argv[4] seems to have a legit value of 7. The original interpreter also depends on mask flags of either 1 or 6, otherwise nothing would get saved.

For the screen top icon bar, argc is also 4 and argv[4] is 0. So it really does not save anything (which does not seem to matter).

Reply:
If I think about it: For the HiRes graphics, it really shouldn't be necessary to save any pixels (since the hires graphics never make it into the engine screen buffers) . Saving the rect dimensions should suffice. So maybe we could always call with mask parameter `0` here and leave argv[4] alone. This needs some testing, though.

Reply:
Ha, wow, that means Sierra never wrote a script that called this correctly. Only KQ6 calls kGraph 15, and they never passed the fifth parameter, so it was always using an undefined value on the stack that happened to work.

Our kernel table enforces this by only allowing calls with just the rect, if there was ever a call with a mask parameter we would have errored, which is why argv[4] will always be undefined:

```
{ SIG_SCI11,          15, MAP_CALL(GraphSaveUpscaledHiresBox), "iiii",                 NULL }, // kq6 hires
```

I guess this code should just be updated to use whatever the value needs to be and include a comment that they never called it right.

Reply:
> I took a closer look, what do you think about this?
> 

Sounds like a good idea. I'll test a bit with a '0' argument (this won't save any pixels, but it will still store the rect coords) and if that works fine, I'll use that...

Reply:
My grand vision has been fulfilled!